File: androidApp/src/main/java/com/mbta/tid/mbta_app/android/favorites/FavoritesView.kt
package com.mbta.tid.mbta_app.android.favorites

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.ButtonDefaults
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.unit.dp
import androidx.lifecycle.compose.LifecycleResumeEffect
import com.mbta.tid.mbta_app.android.R
import com.mbta.tid.mbta_app.android.component.ActionButton
import com.mbta.tid.mbta_app.android.component.ActionButtonKind
import com.mbta.tid.mbta_app.android.component.ErrorBanner
import com.mbta.tid.mbta_app.android.component.NavTextButton
import com.mbta.tid.mbta_app.android.component.SheetHeader
import com.mbta.tid.mbta_app.android.component.routeCard.RouteCardList
import com.mbta.tid.mbta_app.android.util.contrastTranslucent
import com.mbta.tid.mbta_app.android.util.timer
import com.mbta.tid.mbta_app.model.response.AlertsStreamDataResponse
import com.mbta.tid.mbta_app.model.response.GlobalResponse
import com.mbta.tid.mbta_app.model.routeDetailsPage.RouteDetailsContext
import com.mbta.tid.mbta_app.model.routeDetailsPage.RoutePickerPath
import com.mbta.tid.mbta_app.routes.SheetRoutes
import com.mbta.tid.mbta_app.utils.NavigationCallbacks
import com.mbta.tid.mbta_app.viewModel.FavoritesViewModel
import com.mbta.tid.mbta_app.viewModel.IErrorBannerViewModel
import com.mbta.tid.mbta_app.viewModel.IFavoritesViewModel
import com.mbta.tid.mbta_app.viewModel.IToastViewModel
import com.mbta.tid.mbta_app.viewModel.ToastViewModel
import io.github.dellisd.spatialk.geojson.Position
import kotlin.time.Duration.Companion.seconds

@Composable
fun FavoritesView(
    openSheetRoute: (SheetRoutes) -> Unit,
    setLastLocation: (Position) -> Unit,
    setIsTargeting: (Boolean) -> Unit,
    favoritesViewModel: IFavoritesViewModel,
    errorBannerViewModel: IErrorBannerViewModel,
    toastViewModel: IToastViewModel,
    alertData: AlertsStreamDataResponse?,
    globalResponse: GlobalResponse?,
    targetLocation: Position?,
) {
    val now by timer(updateInterval = 5.seconds)
    val state by favoritesViewModel.models.collectAsState()
    val context = LocalContext.current

    fun onAddFavorites() {
        favoritesViewModel.setIsFirstExposureToNewFavorites(false)
        toastViewModel.hideToast()
        openSheetRoute(SheetRoutes.RoutePicker(RoutePickerPath.Root, RouteDetailsContext.Favorites))
    }

    LaunchedEffect(now) { favoritesViewModel.setNow(now) }
    LaunchedEffect(alertData) { favoritesViewModel.setAlerts(alertData) }
    LaunchedEffect(targetLocation) { favoritesViewModel.setLocation(targetLocation) }

    LaunchedEffect(Unit) {
        favoritesViewModel.setContext(FavoritesViewModel.Context.Favorites)
        favoritesViewModel.setActive(active = true, wasSentToBackground = false)
        favoritesViewModel.reloadFavorites()
    }

    LifecycleResumeEffect(Unit) {
        favoritesViewModel.setActive(active = true, wasSentToBackground = false)
        onPauseOrDispose {
            favoritesViewModel.setActive(active = false, wasSentToBackground = true)
        }
    }

    LaunchedEffect(state.awaitingPredictionsAfterBackground) {
        errorBannerViewModel.setIsLoadingWhenPredictionsStale(
            state.awaitingPredictionsAfterBackground
        )
    }

    LaunchedEffect(state.loadedLocation) {
        state.loadedLocation?.let { setLastLocation(it) }
        setIsTargeting(false)
    }
    LaunchedEffect(state.shouldShowFirstTimeToast) {
        if (state.shouldShowFirstTimeToast) {
            toastViewModel.showToast(
                ToastViewModel.Toast(
                    context.getString(R.string.favorite_stops_first_time_toast_message),
                    action =
                        ToastViewModel.ToastAction.Close(
                            onClose = {
                                favoritesViewModel.setIsFirstExposureToNewFavorites(false)
                                toastViewModel.hideToast()
                            }
                        ),
                )
            )
        }
    }

    val routeCardData = state.routeCardData

    Column() {
        SheetHeader(
            title = stringResource(R.string.favorites_link),
            navCallbacks =
                NavigationCallbacks(
                    onBack = null,
                    onClose = null,
                    sheetBackState = NavigationCallbacks.SheetBackState.Hidden,
                ),
            rightActionContents = {
                Row(Modifier, Arrangement.spacedBy(16.dp), Alignment.CenterVertically) {
                    if (!routeCardData.isNullOrEmpty()) {
                        ActionButton(
                            ActionButtonKind.Plus,
                            colors = ButtonDefaults.contrastTranslucent(),
                            action = ::onAddFavorites,
                        )
                        NavTextButton(stringResource(R.string.edit)) {
                            openSheetRoute(SheetRoutes.EditFavorites)
                        }
                    }
                }
            },
        )

        ErrorBanner(errorBannerViewModel, modifier = Modifier.padding(top = 8.dp))
        RouteCardList(
            routeCardData = routeCardData,
            emptyView = {
                NoFavoritesView(::onAddFavorites)
                Spacer(Modifier.weight(1f))
            },
            global = globalResponse,
            now = now,
            isFavorite = { rsd -> (state.favorites?.keys ?: emptySet()).contains(rsd) },
            onOpenStopDetails = { stopId, filter ->
                openSheetRoute(SheetRoutes.StopDetails(stopId, filter, null))
            },
        )
    }
}

File: iosApp/iosApp/Pages/Favorites/FavoritesView.swift
//
//  FavoritesView.swift
//  iosApp
//
//  Created by Melody Horn on 6/12/25.
//  Copyright © 2025 MBTA. All rights reserved.
//

import CoreLocation
import Shared
import SwiftUI

struct FavoritesView: View {
    var errorBannerVM: IErrorBannerViewModel
    var favoritesVM: IFavoritesViewModel
    @State var favoritesVMState: FavoritesViewModel.State = .init()
    @ObservedObject var nearbyVM: NearbyViewModel
    var toastVM: IToastViewModel
    @Binding var location: CLLocationCoordinate2D?

    @State var globalData: GlobalResponse?
    var globalRepository = RepositoryDI().global
    let inspection = Inspection<Self>()
    @State var now = Date.now

    @ScaledMetric private var editButtonHeight: CGFloat = 32

    var body: some View {
        VStack(alignment: .leading, spacing: 0) {
            SheetHeader(
                title: NSLocalizedString("Favorites", comment: "Header for favorites sheet"),
                navCallbacks: .init(onBack: nil, onClose: nil, backButtonPresentation: .floating),
                rightActionContents: {
                    if let routeCardData = favoritesVMState.routeCardData, !routeCardData.isEmpty {
                        ActionButton(kind: .plus, circleColor: Color.translucentContrast, action: { onAddStops() })
                        NavTextButton(
                            string: NSLocalizedString("Edit", comment: "Button text to enter edit favorites flow"),
                            backgroundColor: Color.translucentContrast,
                            textColor: Color.fill2,
                            height: editButtonHeight,
                            width: 64
                        ) {
                            nearbyVM.pushNavEntry(.editFavorites)
                        }
                    }
                }
            ).padding(.bottom, 16)

            ErrorBanner(errorBannerVM, padding: .init([.horizontal, .bottom], 16))
            RouteCardList(
                routeCardData: favoritesVMState.routeCardData,
                emptyView: {
                    NoFavoritesView(
                        onAddStops: {
                            onAddStops()
                        }
                    )
                    .frame(maxWidth: .infinity)
                },
                global: globalData,
                now: now,
                isFavorite: { rsd in favoritesVMState.favorites?.contains(where: { rsd == $0.key }) ?? false },
                pushNavEntry: { nearbyVM.pushNavEntry($0) },
                showStopHeader: true
            )
        }
        .global($globalData, errorKey: "FavoritesView")
        .onAppear {
            favoritesVM.setActive(active: true, wasSentToBackground: false)
            favoritesVM.setAlerts(alerts: nearbyVM.alerts)
            favoritesVM.setContext(context: FavoritesViewModel.ContextFavorites())
            favoritesVM.setLocation(location: location?.positionKt)
            favoritesVM.setNow(now: now.toEasternInstant())
            favoritesVM.reloadFavorites()
        }
        .onReceive(inspection.notice) { inspection.visit(self, $0) }
        .task {
            while !Task.isCancelled {
                now = Date.now
                try? await Task.sleep(for: .seconds(5))
            }
        }
        .task {
            for await model in favoritesVM.models {
                favoritesVMState = model
            }
        }
        .onChange(of: favoritesVMState.awaitingPredictionsAfterBackground) {
            errorBannerVM.setIsLoadingWhenPredictionsStale(isLoading: $0)
        }
        .onChange(of: favoritesVMState.loadedLocation) {
            nearbyVM.lastLoadedLocation = $0?.coordinate
            nearbyVM.isTargeting = false
        }
        .onAppear {
            if favoritesVMState.shouldShowFirstTimeToast {
                showFirstTimeToast()
            }
        }
        .onDisappear {
            toastVM.hideToast()
        }
        .onChange(of: favoritesVMState.shouldShowFirstTimeToast) { shouldShow in
            if shouldShow {
                showFirstTimeToast()
            }
        }
        .onChange(of: nearbyVM.alerts) { favoritesVM.setAlerts(alerts: $0) }
        .onChange(of: location?.positionKt) { favoritesVM.setLocation(location: $0) }
        .onChange(of: now) { favoritesVM.setNow(now: $0.toEasternInstant()) }
        .withScenePhaseHandlers(
            onActive: { favoritesVM.setActive(active: true, wasSentToBackground: false) },
            onInactive: { favoritesVM.setActive(active: false, wasSentToBackground: false) },
            onBackground: { favoritesVM.setActive(active: false, wasSentToBackground: true) }
        )
    }

    func showFirstTimeToast() {
        toastVM.showToast(toast:
            .init(message:
                NSLocalizedString("Favorite stops replaces the prior starred routes feature.",
                                  comment: "Explainer the first time a user sees the new favorites feature"),
                duration: .indefinite,
                isTip: false,
                action: ToastViewModel
                    .ToastActionClose(onClose: { favoritesVM.setIsFirstExposureToNewFavorites(isFirst: false)
                        toastVM.hideToast()
                    })))
    }

    private func onAddStops() {
        favoritesVM.setIsFirstExposureToNewFavorites(isFirst: false)
        toastVM.hideToast()
        nearbyVM.pushNavEntry(
            SheetNavigationStackEntry.routePicker(
                SheetRoutes.RoutePicker(
                    path: RoutePickerPath.Root(),
                    context: RouteDetailsContext.Favorites()
                )
            )
        )
    }
}
